### DISCUSSION TITLE
What could cause runaway queries?

### DISCUSSION THREAD

**Discussion by c0c0n3** [crate/crate#10869] [General]
Hello,

I noticed that every now and then we get runaway queries on our Crate cluster. For example this afternoon,

    select * from sys.jobs

reported that several queries were running since 11am. Most of queries listed there would normally execute in less than a second, so it looks like the query jobs could be deadlocked. For the record, here's one of the queries:

```
SELECT
  time_index AS "time",
  name as "metric",
  1 as violation 
FROM mtekz.etevchargingstation
WHERE detectedviolation = true
union ALL SELECT
  time_index AS "time",
  name as "metric",
  0 as violation 
FROM mtekz.etevchargingstation
WHERE detectedviolation = false
ORDER BY 1,2
```

We have a 3-node K8s-managed deployment with official Crate `4.1.8` Docker images and we noticed that just before those queries started running some of the Crate processes ran out of memory and died, then K8s restarted them. I combed the logs for any clue about why those queries got stuck but had no luck. I don't know if this is relevant at all, but when I killed the query jobs, some `kill` statements reported that two rows had been deleted even though `select * sys.jobs` had shown one row for that job ID, for example

    kill 'a76e36c2-5a56-42db-880f-874914e5ef50';
    OK, 2 rows affected (0.005 sec)

Any idea what could cause this and how to prevent it from happening? Is there a way to automatically kill jobs that have been running for hours? Also, are those runaway queries tied to actual objects on the Java heap so the GC won't be able to reclaim memory until the corresponding job is killed?

Thanks alot!!

**Comment by eazydripz** [comment]
Thank you.

**Comment by hlcianfagna** [comment]
Hi,

>  Is there a way to automatically kill jobs that have been running for hours?

One way to do this is with the `statement_timeout` session setting 
https://crate.io/docs/crate/reference/en/5.4/config/session.html#conf-session-statement-timeout
which was introduced in CrateDB 5.4.0
